src/EVM/SMT.hs

```diff
@@ -811,6 +787,117 @@ getVars parseFn inst names = Map.mapKeys parseFn <$> foldM getOne mempty names
+            slot' <- queryValue slot
+            pure $ addElem addr' slot' m fun) Map.empty sreads
+
+  where
+
+    addElem :: W256 -> W256 -> Map W256 (Map W256 W256) -> (W256 -> W256 -> W256) -> Map W256 (Map W256 W256)
+    addElem addr slot store fun =
+      case Map.lookup addr store of
+        Just m -> Map.insert addr (Map.insert slot (fun addr slot) m) store
+        Nothing -> Map.insert addr (Map.singleton slot (fun addr slot)) store
+
+
+    queryValue :: Expr EWord -> IO W256
+    queryValue (Lit w) = pure w
+    queryValue w = do
+      let expr = toLazyText $ exprToSMT w
+      raw <- getValue inst expr
+      case parseCommentFreeFileMsg getValueRes (T.toStrict raw) of
+        Right (ResSpecific (valParsed :| [])) ->
+          case valParsed of
+            (_, TermSpecConstant sc) -> pure $ parseW256 sc
+            _ -> error "Internal Error: cannot parse model for storage index"
+        r -> parseErr r
+
+
+is1DArrConst :: QualIdentifier -> Bool
+is1DArrConst = \case
+  Qualified (IdSymbol "const") sort -> is1DArray sort
+  _ -> False
+
+is1DArray :: Sort -> Bool
+is1DArray = \case
+  SortParameter (IdSymbol "Array") (sort1 :| [sort2]) ->
+    isBitVec256 sort1 && isBitVec256 sort2
+  _ -> False
+
+is2DArrConst :: QualIdentifier -> Bool
+is2DArrConst = \case
+  Qualified (IdSymbol "const") sort -> is2DArray sort
+  _ -> False
+
+is2DArray :: Sort -> Bool
+is2DArray = \case
+  SortParameter (IdSymbol "Array") (sort1 :| [sort2]) ->
+    isBitVec256 sort1 && is1DArray sort2
+  _ -> False
+
+isBitVec256 :: Sort -> Bool
+isBitVec256 t =
+  t == SortSymbol (IdIndexed "BitVec" (IxNumeral "256" :| []))
+
+isStore :: QualIdentifier -> Bool
+isStore t = t == Unqualified (IdSymbol "store")
+
+
+-- | Interpret an 1-dimentional array as a function
+interpret1DArray :: (Map Symbol Term) -> Term -> (W256 -> W256)
+interpret1DArray env = \case
+  -- variable reference
+  TermQualIdentifier (Unqualified (IdSymbol s)) ->
+    case Map.lookup s env of
+      Just t -> interpret1DArray env t
+      Nothing -> error "Internal error: unknown identifier, cannot parse array"
+  -- (let (x t') t)
+  TermLet (VarBinding x t' :| []) t -> interpret1DArray (Map.insert x t' env) t
+  TermLet (VarBinding x t' :| lets) t -> interpret1DArray (Map.insert x t' env) (TermLet (NonEmpty.fromList lets) t)
+  -- (as const (Array (_ BitVec 256) (_ BitVec 256))) SpecConstant
+  TermApplication asconst (TermSpecConstant val :| []) | is1DArrConst asconst ->
+    \_ -> parseW256 val
+  -- (store arr ind val)
+  TermApplication store (arr :| [TermSpecConstant ind, TermSpecConstant val]) | isStore store ->
+    \x -> if x == parseW256 ind then parseW256 val else interpret1DArray env arr x
+  t -> error $ "Internal error: cannot parse array value. Unexpected term: " <> (show t)
+
+-- | Interpret a 2-dimentional array as a function
+interpret2DArray :: (Map Symbol Term) -> Term -> (W256 -> W256 -> W256)
+interpret2DArray env = \case
```

Comment:
There is some code duplication between `interpret2DArray` and `interpret1DArray`, but in order to generalise to `interpretNDArray` we would need a dependent types for the result, and I'm not sure if it's worth the hustle (though I'd be somewhat curious to try it out).